hudi-io/src/main/java/org/apache/hudi/storage/HoodieStorage.java

```diff
@@ -309,11 +309,15 @@ public final StorageConfiguration<?> getConf() {
   @PublicAPIMethod(maturity = ApiMaturityLevel.EVOLVING)
   public final void createImmutableFileInPath(StoragePath path,
                                               Option<byte[]> content) throws HoodieIOException {
+    createImmutableFileInPath(path, content, needCreateTempFile());
+  }
+
+  @PublicAPIMethod(maturity = ApiMaturityLevel.EVOLVING)
+  public final void createImmutableFileInPath(StoragePath path,
+                                              Option<byte[]> content, boolean needTempFile) throws HoodieIOException {
     OutputStream fsout = null;
     StoragePath tmpPath = null;
 
-    boolean needTempFile = needCreateTempFile();
```

Comment:
why we need an explicit flag for this?

Reply:
> why we need an explicit flag for this?

`HoodieStorage::needCreateTempFile` only be true for HDFS, but for other fs, such as local fs, we should also avoid  visiting the intermediate state of the file. So completely, regardless of the underlying file system, we will always use the temp file method to create files in this scenario.

Reply:
We have assumption that the storage has two categories: HDFS and Object Store, the later should always be atomic for file creation so that the renaming is only needful for HDFS.

If you think there is other fs system we need to support, at least to fix the `HoodieStorage::needCreateTempFile` itself.

Reply:
> We have assumption that the storage has two categories: HDFS and Object Store, the later should always be atomic for file creation so that the renaming is only needful for HDFS.
> 
> If you think there is other fs system we need to support, at least to fix the `HoodieStorage::needCreateTempFile` itself.

I fix `HoodieStorage::needCreateTempFile` to for all storage without write-transaction should create temp file.

Reply:
We can not do this, the flag `isWriteTransactional` does not set up right for many object stores. Let's just keep it as it is, local fs is only for testing purpose so we should be good.

Reply:
> We can not do this, the flag `isWriteTransactional` does not set up right for many object stores. Let's just keep it as it is, local fs is only for testing purpose so we should be good.

Got it! But can I just add a judgment of whether it's local fs or not? Because ut's tests rely on local fs, if we do not add this judgment, ut will have unexpected logic.

Reply:
Let's not change this logic.

Reply:
> Let's not change this logic.

Ok, I have restored the original logic~